invader_payment_paypal/services/payment_paypal.py

```diff
@@ -0,0 +1,232 @@
+
+PAYPAL_ERROR = {"invalid_client": _("Invalid Paypal Credential")}
+
+PAYPAL_STATUS = {
+    "CREATED": "draft",
+    "SAVED": "draft",
+    "APPROVED": "pending",
+    "VOIDED": "error",
+    "COMPLETED": "done",
+}
+
+
+class PaymentServicePaypal(AbstractComponent):
+
+    _name = "payment.service.paypal"
+    _inherit = "base.rest.service"
+    _usage = "payment_paypal"
+    _description = "REST Services for Paypal payments"
+
+    @property
+    def payment_service(self):
+        return self.component(usage="invader.payment")
+
+    def _validator_checkout_order(self):
+        """
+        Validator of confirm_payment service
+        target: see _allowed_payment_target()
+        payment_mode_id: The payment mode used to pay
+        :return: dict
+        """
+        res = self.payment_service._invader_get_target_validator()
+        res.update(
+            {
+                "payment_mode_id": {
+                    "coerce": to_int,
+                    "type": "integer",
+                    "required": True,
+                },
+                "return_url": {"type": "string", "required": True},
+                "cancel_url": {"type": "string", "required": True},
+            }
+        )
+        return res
+
+    def _validator_return_checkout_order(self):
+        schema = {"redirect_to": {"type": "string"}}
+        return Validator(schema, allow_unknown=True)
+
+    def _get_api_url(self, acquirer, path):
+        if acquirer.environment == "prod":
+            return "https://api.paypal.com/{}".format(path)
+        else:
+            return "https://api.sandbox.paypal.com/{}".format(path)
+
+    def _catch_paypal_error(self, response):
+        data = response.json()
+        if data:
+            if data.get("error") in PAYPAL_ERROR:
+                raise UserError(PAYPAL_ERROR[data["error"]])
+            if data.get("error_description"):
+                raise UserError(data["error_description"])
+            if data.get("message"):
+                raise UserError(data["message"])
+        raise UserError(response.content)
+
+    def _get_api_token(self, acquirer):
+        response = requests.post(
+            self._get_api_url(acquirer, "v1/oauth2/token"),
+            headers={"Accept": "application/json"},
+            auth=(acquirer.paypal_client_id, acquirer.paypal_secret),
+            params={"grant_type": "client_credentials"},
+        )
+        if response.status_code != 200:
+            self._catch_paypal_error(response)
+        data = response.json()
+        return "{} {}".format(data["token_type"], data["access_token"])
+
+    def _create_paypal_order(self, acquirer, data):
+        token = self._get_api_token(acquirer)
+        response = requests.post(
+            self._get_api_url(acquirer, "v2/checkout/orders"),
+            headers={
+                "Content-Type": "application/json",
+                "Authorization": token,
+            },
+            json=data,
+        )
```

Comment:
@sebastienbeau We should log rqst and response of calls to the API. it's very important for debug purpose or in case of trouble in production